### DISCUSSION TITLE
prefetch count and batch processing EventingBasicConsumer

### DISCUSSION THREAD

**Discussion by minhhungit** [rabbitmq/rabbitmq-dotnet-client#1237] [Q&A] [answered]
I'm using prefetch count (for example 10) with `EventingBasicConsumer`, it return message one by one in **Received** event. how can I  receive all 100 messages in **one times**. 

If I hold messages I got in a safe thread LIST and count it, I also CAN NOT know exact when rabbitmq return enough messages for me in case my queue has less message than prefetch count ( for example queue has only 7 messages left)

Is there a way to implement a "timeout" for the prefetch, I mean it might work if I set "fetch time" for it, so I just wait until I get enough 10 messages or I get timeout

**Comment by michaelklishin** [comment, marked as answer]
In all messaging protocols RabbitMQ supports, deliveries happen one by one. This is reflected in the consumer handler API signature.

You can either batch the deliveries or consider using streams and the RabbitMQ stream client which allows you to work with stream offsets (a sliding window of positions).

**Reply by minhhungit** [reply]
thank you, I will try to check it

**Reply by boskjoett** [reply]
You said "consider using streams and the RabbitMQ stream client". What do you mean by this?  Where can I find more info about using streams and the RabbitMQ stream client?

**Reply by michaelklishin** [reply]
[In the docs](https://rabbitmq.com/streams.html)
